app/models/exchange_rate/provided.rb

```diff
@@ -6,12 +6,31 @@ module ExchangeRate::Provided
   class_methods do
     private
 
-      def fetch_rates_from_provider(from:, to:, dates:, cache: false)
+      def fetch_rates_from_provider(from:, to:, start_date:, end_date: Date.current, cache: false)
         return [] unless exchange_rates_provider.present?
 
-        dates.map do |date|
-          fetch_rate_from_provider from:, to:, date:, cache:
-        end.compact
+        response = exchange_rates_provider.fetch_exchange_rate_for_date_range \
+          from: from,
+          to: to,
+          date_start: start_date,
+          date_end: end_date
+
+        if response.success?
+          response.rates.map do |exchange_rate|
+            rate = ExchangeRate.new \
+              from_currency: from,
+              to_currency: to,
+              date: exchange_rate.dig(:date).to_date,
+              rate: exchange_rate.dig(:rate)
+
+            rate.save! if cache
+            rate
+          rescue ActiveRecord::RecordNotUnique
+            next
```

Comment:
I think this should fix #1037 as well correct?

Reply:
Yes!